### DISCUSSION TITLE
What I could not do with Toga

### DISCUSSION THREAD

**Discussion by annunal** [beeware/toga#2345] [Q&A]
First of all let me make my congratulations to the development team that allowed me to create my first mobile application using python that is my development language. I am new in Toga, only few days but was able to obtain something already useful.

However I stil face a number of problems that I could not solve through searching and googling:

- **An image on a button**. I would like to have a clickable image or icon.  All the tests I did do no t work properly.  I could add icon only to Commands that however are staying always on top of the main window

- **an image in a Table.** The table is very simple but it does not allow to format each cell or include in a cell a nother object, like an image or a bold text etc

- the **push notification** that I understand is still in early development stage.

I saw that a number of requests have been done for the first point but I do not know if in the meantime it was solved. If someone has been able to solve the first or the second point, I would be extremely grateful to see their coding .

**Comment by freakboy3742** [comment]
Icons on buttons have been implemented recently (#2310); it's available now in the main branch, and will be formally released as part of Toga 0.4.2.

Adding complex content to Table, like images or rich-formatted text, isn't a simple feature request. In many cases, we're going to be constrained by the underlying platforms, which don't always expose a rich API on their table widgets. I'm not fundamentally opposed to the idea of supporting "richer" content types in Table; but it's not something that I would expect to be implemented any time soon, unless someone else contributes a patch.

Push notifications are still in the design phase (see #907 and #1890).

**Comment by annunal** [comment]
Thanks Russel for the updates.

Is there an example of image on button I can use ?   And, if the release was done recently, how I can update  my system to benefit from these updates ?  using  
pip install toga --upgrade

the version 0.4.1 is installed, not yet the 0.4.2.

About the table, maybe starting from icons in a cell instead of 'any possible widget'. could be a possibility ?

**Reply by freakboy3742** [reply]
The Button example in the main branch [contains an example of an Icon button](https://github.com/beeware/toga/blob/main/examples/button/button/app.py#L72).

To use the main branch, you need to install from GitHub, not from PyPI. Instead of installing `toga-core==0.4.1` and `toga-cocoa==0.4.1` (or whatever backend you're using), you specify `toga-core@git+https://github.com/beeware/toga.git@<hash>#subdirectory=core` and `toga-cocoa@git+https://github.com/beeware/toga.git@<hash>#subdirectory=cocoa` (replacing `<hash>` in those commands with a specific Git commit hash - `61dff56` would be the value at the time I'm writing this). 

As for images in tables - Tables *do* support *icons* in cells. In Toga's terminology, an Icon and an Image are different.

If you specify the value of a table as `(icon, text)` or `(icon, None)`, Toga will put the icon in the table cell. However, on Windows, icons are only supported in the first column (this is a limitation of Windows); and on Android, table icons aren't supported at all at present. See the [documentation on Table](https://toga.readthedocs.io/en/stable/reference/api/widgets/table.html) for details.

The Android table implementation is also beta at present, and will be significantly altered (hopefully in the not-too-distant future).  This will also include adding a Table implementation on iOS, and an implementation of Tree on both iOS and Android. Beyond that, the core team have no plans to revisit Table functionality. However, we're open to contributions from people outside the core team that add features to the Table API.

**Comment by annunal** [comment]
Dear Russel,

thanks for the hints. I was able to run a version with a clickable button and an icon
![image](https://github.com/beeware/toga/assets/10267112/f92d0a97-4ae4-45dc-adb1-691ba248abcd)

However,  when I launch the command 
briefcase run android

the system continues to use the old version 0.4.1 and so it does not recognize the Icon attribute of the Button.  What chabges I need to do ?    Thanks

**Reply by freakboy3742** [reply]
The version packages you have in your local environment won't impact the versions that are in a packaged app. If you want your Android app to have the update packages, you need to modify the `pyproject.toml` file to use the git references, rather than a version number.
